srcDoc/custom_types.rst

```diff
@@ -0,0 +1,578 @@
+---------------------------
+
+A string at the beginning of the execution should indicate that you are using a
+custom type and that this is not a regular execution.
+
+
+Custom Types + `FlatData`
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+*FlatData* types are at this point in *RTI Perftest* associated to regular types.
+If you want to use *FlatData* with for your Custom Types, follow these
+additional steps:
+
+- Create a copy of your original type (which should not be `FlatData`) in the same IDL file.
+- Make this new type mutable (``@mutable``) and make it FlatData compatible (``@language_binding(FLAT_DATA)``).
+- Implement the additional API custom type functions of customtype.cxx for *FlatData* types.
+  This is a *must*, as the type is ``@mutable``, and every field is considered as optional
+  and not sent into the wire if it is not set.
+- Run the build script with the command-line parameter ``--customType <type>`` and ``--customTypeFlatData <flat_type>``.
+- Run *RTI Perftest* as usual.
+
+Full example using Custom Types
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The following types will be used for this example. The code can be found under
+``resource/examples_customType/no_key_large``.
+
+.. code-block:: c
+
+    const long SIZE_TEST_SEQ = 100;
+    const long SIZE_TEST_STRING = 128;
+
+    enum TestEnum {
+        ENUM1,
+        ENUM2
+    };//@Extensibility FINAL_EXTENSIBILITY
+
+    struct StringTest {
+        string<SIZE_TEST_STRING> test_string;
+    };//@Extensibility FINAL_EXTENSIBILITY
+
+    struct SeqTest {
+        sequence<long, SIZE_TEST_SEQ> test_seq;
+    };//@Extensibility FINAL_EXTENSIBILITY
+
+    struct Test {
+        long test_long;
+        TestEnum test_enum;
+        StringTest test_string;
+        SeqTest test_seq;
+    };//@Extensibility FINAL_EXTENSIBILITY
+
+These functions in the example are spreaded across several `idl` files to show the
+multiple files capability of the feature.
+
+These are the steps needed to use the above type in *RTI Perftest* for the
+C++ (Traditional) API:
+
+1. Copy all the `idl` files into the `srcIdl/customType/` folder.
+
+2. The following functions should be implemented (optionally) to properly
+   initialize and set the custom type structures.
+
+    - **initialize_custom_type_data**
+    This function is used to initialize your data.
+    Using this function, you will be able to allocate memory or set an immutable
+    field of the data.
+    The function takes one argument:
+
+        - A reference to custom type data.
+
+    .. code-block:: c
+
+        bool initialize_custom_type_data(RTI_CUSTOM_TYPE &data)
+        {
+            bool success = true;
+            if (!data.test_seq.test_seq.ensure_length(SIZE_TEST_SEQ, SIZE_TEST_SEQ)) {
+                success = false;
+            }
+            data.test_enum = ENUM1;
+            return success;
+        }
+
+    - **register_custom_type_data**
+    This function is used to set your data before being registered. It is only
```

Comment:
I'm not sure I understand the use and thus requirements for this function.  

So this is only needed for tests using keys, which would use the TestDataKeyed_t (and variants)?

So, for a keyed test, the test will create keys in the form of unsigned long values that this function will need to store into the key fields defined by the user data type if it is defined with a key field(s)?  Is that the idea?  But why is this needed?  Since the octet key member is still a key member, couldn't the test framework just continue to set that value itself without requiring setting of any key members inside the user structure and still create the number of instances required for the test?

In short, I don't think this function is required to support the testing of keys when using user data types...